docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+invalid.
+
+#### `automerge_type`
+
+This can take the value `"text"` if  placed on `string` properties to indicate
+that they should be represented in an `Automerge.Text` data type.
+
+
+### Strange Perspectives
+
+This model introduces some counter-intuitive properties. For example, I might
+"create an issue" in a repository and anyone who is tracking me would see that
+issue, but people who are tracking the project but don't have me in their
+tracking graph will only see the issue if the maintainer replies to it. It's
+hard to see how you would do things like "link to an issue" under these
+constraints. This is inherent to the network model though, rather than being a
+specific problem of this architecture.
+
+We can work around some of this weirdness using seed nodes. If we consider
+seed nodes.
+
+## APIs
+
+The APIs librad will provide:
+
+- enumerate collaborative objects of a particular type
+- retrieve an object with a particular ID as a JSON representation for
+  applications which only wish to read data
+- retrieve an object with a particular ID as an Automerge document for
+  applications which wish to write data
+- update an object by providing the bytes of an automerge change which updates
+  the document
+- create a new object from a JSON object, a JSON schema, and a type name
+  
+Note that I am referring to "the binary representation of an automerge x" 
+because the automerge API works in terms of binary changes.
+
+This new API will live in a new top-level module at
+`librad::collaborative_objects`. An initial sketch looks like this:
+
+```rust
+struct CollaborativeObjectStore {
+    storage: git::storage::Pool,
+    signer: signer::Signer,
+}
+
+enum History {
+    Automerge(Vec<Vec<u8>>)
+}
+
+struct ObjectId(String);
+struct TypeName(String);
+struct Schema(..);
+
+struct CollaborativeObject {
+    project_urn: Urn,
+    typename: TypeName,
+    schema: Schema,
+    id: ObjectId,
+    author: Person, 
+    json: serde_json::Value,
+    history: History, 
+}
+
+struct NewObjectSpec {
+    project_urn: Urn,
+    typename: TypeName,
+    history: History,
+    schema_json: serde_json::Value,
+}
+
+impl CollaborativeObjectStore {
+    fn retrieve_objects(&self, project_urn: Urn, typename: TypeName) -> Result<Vec<CollaborativeObject>, _>
+    fn retrieve_object(&self, project_urn: Urn, typename: TypeName, id: ObjectId) -> Result<CollaborativeObject, _>
+    fn create_object(&self, spec: NewObjectSpec) -> Result<CollaborativeObject, _>
+    fn update_object(&self, project_urn: Urn, id: TypeName, changes: History) -> Result<CollaborativeObject, _>
+}
+```
+
+
+## Blessed Data Types
+
+This project metadata mechanism is extremely broad, which has a lot of upsides
+but it runs the risk of running into XMPP style extension hell, where every
+peer is running a different set of extensions. It might be worthwhile to bundle
+a few core extensions with librad - issues for example.
+
+## Alternative Approaches
```

Comment:
I would love to see a comparison with the event log style approach of [SSB](https://ssbc.github.io/scuttlebutt-protocol-guide/#feeds) and [DAT](https://datprotocol.github.io/how-dat-works).

Reply:
This is actually subsumed by “Domain Specific CRDTs” and “Blessed Datatypes” — SSB applies an arbitrary fold function to a fixed set of datatypes, and leaves it to applications to agree on what the function is. The prevalent “multiwriter” thing for DAT is, afaik, “hypermerge” (which, incidentally, is based on automerge).

Reply:
> The prevalent “multiwriter” thing for DAT is, afaik, “hypermerge” (which, incidentally, is based on automerge).

Thanks for clarifying. I’m not intimately familiar with DAT.


> This is actually subsumed by “Domain Specific CRDTs”

I read this section as being more focused on CRDTs and canoncial representations where as SSB is a bit more loose with it. I’ll add some more specific comments there.